### ISSUE TITLE
Confusing "the parameter type `Self` may not live long enough" error with dyn HRTB callback taking GAT

### ISSUE THREAD

**Issue by Kestrer** [rust-lang/rust#95331]
I tried this code:

```rust
#![feature(generic_associated_types)]

trait FooBad {
    type Item<'a>;
    fn items(walker: &mut dyn for<'a> FnMut(Self::Item<'a>));
}
```

I expected to see this happen: No error.

Instead, this happened:

```
error[E0311]: the parameter type `Self` may not live long enough
 --> src/main.rs:5:22
  |
5 |     fn items(walker: &mut dyn for<'a> FnMut(Self::Item<'a>));
  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: the parameter type `Self` must be valid for the anonymous lifetime defined here...
 --> src/main.rs:5:22
  |
5 |     fn items(walker: &mut dyn for<'a> FnMut(Self::Item<'a>));
  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...so that the reference type `&mut dyn for<'a> FnMut(<Self as FooBad>::Item<'a>)` does not outlive the data it points at
 --> src/main.rs:5:22
  |
5 |     fn items(walker: &mut dyn for<'a> FnMut(Self::Item<'a>));
  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The strange thing is that this alternative version, which should in theory be totally identical, works fine:

```rs
trait FooGood {
    type Item<'a>;
    fn items(walker: &mut dyn ItemWalker<Self>);
}

trait ItemWalker<T: ?Sized + FooGood>: for<'a> FnMut(T::Item<'a>) {}
impl<T: ?Sized + FooGood, F: ?Sized> ItemWalker<T> for F where F: for<'a> FnMut(T::Item<'a>) {}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.61.0-nightly (d53246fed 2022-03-25)
binary: rustc
commit-hash: d53246fedde4c193eae8a003546a8f0f9f85d223
commit-date: 2022-03-25
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by nikomatsakis** [comment]
This is a problem with for and the lack of implied bounds (which I would like to fix), but I think it is fairly orthogonal from GATs.

What is happening here is that the `&'b mut dyn for<'a> FnMut(Self::Item<'a>)` is only well formed if `dyn for<'a> FnMut(...): 'b`. Normally, we would add an implied bound that indicates that this is the case, but because of the higher-ranked region `'a` we don't do so, I think, and then we wind up trying to prove that `Self::Item<'a>: 'b`, which we can do if we can show that `Self: 'b`, but we can't (normally the implied bound would let us do so).

The same problem arises without GATs in a case like this:

```rust
trait FooBad: for<'a> TheItem<'a>  {
    fn items(walker: &mut dyn for<'a> FnMut(<Self as TheItem<'a>>::Item));
}

trait TheItem<'a> {
    type Item;
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f401714767405dc5a686cd61c9984e15

**Comment by aliemjay** [comment]
A workaround:

```rust
#![feature(generic_associated_types)]

trait FooBad {
    type Item<'a>;
    fn items<'b>(walker: &'b mut dyn for<'a> FnMut(Self::Item<'a>))
    where
        Self: 'b;
}

// or, if the method recieves `&self`, make use of the implied bound `Self: 'b`
trait FooBadWithSelf {
    type Item<'a>;
    fn items<'b>(&'b self, walker: &'b mut dyn for<'a> FnMut(Self::Item<'a>));
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=442fcaf520e2f3fdfb2b515ae2f791bb

**Comment by aliemjay** [comment]
Another manifestation of the same problem:
```rust
#![feature(generic_associated_types)]
trait Foo { type Item<'a>; }

// this has an implicit bound `S: 's`.
struct MyFnGood<'s, S: Foo>(&'s dyn Fn(S));

// requires an explicit bound `S: 's`.
struct MyFnBad<'s, S: Foo>(&'s dyn for<'a> Fn(S::Item<'a>));
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=99ec21946e2afbb7926422c658917d92
